docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,750 @@
+
+        + :code:`sync_state: gt4py.storage.SyncState`. If `managed="gt4py"` indicates which of the
+          provided buffers, `data` or `device_data`, is up to date at the time of initialization.
+
+        For common keyword-only arguments, please see below.
+
+    If :code:`copy=False` and neither :code:`data` nor :code:`device_data` are provided, the other
+    arguments are used to allocate an appropriate buffer without initialization (equivalent to call
+    :code:`empty()`). If :code:`data` or :code:`device_data` is provided, the consistency of the
+    parameters with the buffers is validated.
+
+Optional Keyword-Only Parameters
+================================
+
+Additionally, these **optional** keyword-only parameters are accepted:
+
+:code:`aligned_index: Sequence[int]`
+    The index of the grid point to which the memory is aligned. Note that this only partly takes the
+    role of the former :code:`default_origin` parameter, since it does not imply anything about the
+    origin or domain when passed to a stencil. It defaults to the lower indices of the
+    :code:`halo` parameter.
+
+:code:`alignment_size: Optional[int]`
+    The buffers are allocated such that :code:`mod(aligned_addr, alignment_size) == 0`, where
+    :code:`aligned_addr` is the memory address of the grid point denoted by :code:`aligned_index`.
+
+    It defaults to :code:`1`, which indicates no alignment.
+
+:code:`defaults: Optional[str]`
+    It can be used in the way of the current :code:`backend` parameter. For each backend, as well
+    as for the keys :code:`"F"` and :code:`"C"` (equivalent to the same values in the :code:`order`
+    parameter for NumPy allocation routines) a preset of suitable parameters is provided. Explicit
+    definitions of additional parameters are possible and they override its default value from the
+    preset.
+
+:code:`device: Optional[str]`
+    Indicates whether the storage should contain a buffer on an accelerator device. Currently it
+    only accepts :code:`"gpu"` or :code:`None`. Defaults to :code:`None`.
+
+:code:`dims: Optional[Sequence[str]`
+    Sequence indicating the semantic meaning of the dimensions of this storage. This is used to
+    determine the default layout for the storage. Currently supported will be :code:`"I"`,
+    :code:`"J"`, :code:`"K"` and additional dimensions as string representations of integers,
+    starting at :code:`"0"`.
+
+:code:`halo: Optional[Sequence[Union[int, Tuple[int, int]]]`
+    Sequence of length :code:`ndim` where each entry is either an :code:`int` or a 2-tuple
+    of :code:`int` s. A sequence of integer numbers represent a symmetric halo with the specific
+    size per dimension, while a sequence of 2-tuple specifies the start and end boundary sizes on
+    the respective dimension, which can be used to denote asymmetric halos. It defaults to no halo,
+    i.e. :code:`(0, 0, 0)`. (See also Section :ref:`domain_and_halo`)
+
+:code:`layout: Optional[Sequence[int]]`
```

Comment:
nit: `layout` is a slightly overloaded term, we're using it already to indicate the number of ranks along edges of a tile face. Could this be named `stride_order` or something similar?

Reply:
When using `as_storage`, is this derived by default from the array passed as input? If so, can you describe that here? Also, what is the interplay between this and the `dims` argument, if only one is given?